File: Data Structures & Algorithms/subsets/submission-4.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res  = []
        def dfs(i, ar):
            if i == len(nums):
                res.append(ar.copy())
                return 
            ar.append(nums[i])
            print(ar)
            dfs(i+1, ar)
            ar.remove(nums[i])
            dfs(i+1, ar)
        dfs(0, [])
        return res


        



        

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.java
class Solution {
    List<Integer> memo = new ArrayList<Integer>();
    Integer max = Integer.MIN_VALUE;
    private int dfs(List<Integer> nums, Integer i) {
        
        if(i == nums.size()-1) {
            return 1;
        }

        int result = 1;
        for(int j=i+1; j<nums.size(); j++) {
            if(nums.get(i) < nums.get(j)) {
                System.out.println(nums.get(i)+" "+nums.get(j));
                int sresult = 1+dfs(nums, j);
               
                result = Math.max(result, sresult);
                 System.out.println("result"+ result);
            } 
        }
        max = Math.max(max, result);
        return result;
    }
    public int lengthOfLIS(int[] nums) {
        List<Integer> list = Arrays.stream(nums)
                        .boxed()
                        .collect(Collectors.toList());

        for(int j=0; j<list.size(); j++) {
            dfs(list, j);
        }
        
        return (max == -2147483648) ? 1 : max;
    }   
}
